File: auth/src/iosMain/kotlin/com/well/auth/credentialProviders/GoogleProvider.kt
package com.well.auth.credentialProviders

import cocoapods.GoogleSignIn.*
import com.well.auth.Context
import com.well.auth.LoginView
import com.well.utils.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine
import platform.Foundation.NSError
import platform.Foundation.NSURL
import platform.UIKit.UIApplication
import platform.darwin.NSObject
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

actual class GoogleProvider actual constructor(context: Context) : CredentialProvider(context) {
    private val signIn = GIDSignIn.sharedInstance()!!.apply {
        clientID = "567261840338-fbpj843t6o0lkscvafv7l4368es1h5ru.apps.googleusercontent.com"
    }

    actual override suspend fun getCredentials(loginView: LoginView): AuthCredential =
        suspendCancellableCoroutine { continuation ->
            signIn.presentingViewController = loginView
            signIn.delegate = object : NSObject(), GIDSignInDelegateProtocol {
                override fun signIn(
                    signIn: GIDSignIn?,
                    didSignInForUser: GIDGoogleUser?,
                    withError: NSError?
                ) {
                    signIn?.delegate = null
                    val token = didSignInForUser?.authentication?.idToken
                    when {
                        withError?.isGoogleSignInCanceled == true -> continuation.cancel()
                        withError != null -> continuation.resumeWithException(withError)
                        token != null -> continuation.resume(AuthCredential.GoogleCredential(token))
                        else -> continuation.resumeWithException(IllegalStateException())
                    }
                }

                private val NSError.isGoogleSignInCanceled
                    get() = domain == kGIDSignInErrorDomain && code == kGIDSignInErrorCodeCanceled
            }
            signIn.signIn()
        }

    override fun application(app: UIApplication, openURL: NSURL, options: Map<Any?, *>): Boolean =
        signIn.handleURL(openURL)
}

File: iosApp/Well/SharedPodImplementation/GoogleProvider.swift
//
//  GoogleProvider.swift
//  Well
//
//  Created by Philip Dukhov on 2/14/21.
//  Copyright © 2021 Well. All rights reserved.
//

import SharedMobile
import GoogleSignIn

final class GoogleProvider: CredentialProvider {
    private let systemContext: SystemContext

    override init(systemContext: SystemContext) {
        self.systemContext = systemContext
        super.init(systemContext: systemContext)
    }

    override func getCredentials(completionHandler: @escaping (AuthCredential?, Error?) -> Void) {
        GIDSignIn.sharedInstance.signIn(
            with: GIDConfiguration(
                clientID: (Bundle.main.object(forInfoDictionaryKey: "firebaseInfo") as! [String: String])["clientId"]!
            ),
            presenting: systemContext.rootController
        ) { user, error in
            if let token = user?.authentication.idToken {
                completionHandler(AuthCredential.GoogleCredential(token: token), nil)
                return
            }
            guard let error = error as? GIDSignInError else { fatalError("No token and no error \(String(describing: error))") }
            if error.code == .canceled {
                completionHandler(nil, KotlinCancellationException().toNSError())
            } else {
                completionHandler(nil, error)
            }
        }
    }

    override func application(app: UIApplication, openURL: URL, options: [AnyHashable: Any] = [:]) -> Bool {
        GIDSignIn.sharedInstance.handle(openURL)
    }
}
